modules/gdscript/gdscript.cpp

```diff
@@ -2178,6 +2179,16 @@ void GDScriptLanguage::finish() {
 			scr->clear();
 		}
 		s = s->next();
+		if (!s) {
+			break;
+		}
+
+		// Destructing scr might lead to s being released. So we store a weakref here to check later
+		wnext->set_obj(s->self());
+		scr.unref();
```

Comment:
`unref()` only leads to more problems.

Here's how you could do this without `unref()`:
```cpp
void GDScriptLanguage::finish() {
	_call_stack.free();

	// Clear the cache before parsing the script_list
	GDScriptCache::clear();

	// Clear dependencies between scripts, to ensure cyclic references are broken
	// (to avoid leaks at exit).
	SelfList<GDScript> *s = script_list.first();
	while (s) {
		{
			// This ensures the current script is not released before we can check
			// what's the next one in the list (we can't get the next upfront because we
			// don't know if the reference breaking will cause it -or any other after
			// it, for that matter- to be released so the next one is not the same as
			// before).
			Ref<GDScript> scr = s->self();
			script_list.remove(s);

			if (scr.is_valid()) {
				for (KeyValue<StringName, GDScriptFunction *> &E : scr->member_functions) {
					GDScriptFunction *func = E.value;
					for (int i = 0; i < func->argument_types.size(); i++) {
						func->argument_types.write[i].script_type_ref = Ref<Script>();
					}
					func->return_type.script_type_ref = Ref<Script>();
				}
				for (KeyValue<StringName, GDScript::MemberInfo> &E : scr->member_indices) {
					E.value.data_type.script_type_ref = Ref<Script>();
				}

				// Clear backup for scripts that could slip out of the cyclic reference
				// check
				scr->clear();
			}
		}

		s = script_list.first();
		if (!s) {
			break;
		}
	}
	script_list.clear();
	function_list.clear();
}
```

Reply:
I did think of doing it like this first. However I wasn't sure that it was guaranteed that `first` would always update. I.e. if there is a second script that has a ref to `script_list.first()` I don't think this would work but I might be wrong.

Reply:
Well, I just realized that the `if (!s) { break; }` at the end is not needed, as there's the `while (s)` check.

> I did think of doing it like this first. However I wasn't sure that it was guaranteed that first would always update.

With the scope, it's guaranteed that `scr` will be cleaned up. So I think it's as safe as your first version. Could you test if you have issues with my suggestion? Or do you have a MRP so I could test it on my side?

Reply:
I can test in a bit. Your scope does not guarantee `scr` is cleaned up though. If there is a ref to `scr` from a different script that has not yet been freed it will still exist after the local `Ref` is destructed. The question is if that can happen or not.

Reply:
Yes as I feared. Your solution produces an infinite loop because `scr` is indeed not freed and `script_list.first() == scr`.